spec/function.dd

```diff
@@ -173,124 +173,319 @@ int foo();
+same arguments.)
 
-    $(P A pure function that accepts only parameters without mutable
-    indirections and returns a result that has mutable indirections is called a
-    $(I pure factory function). An implementation may assume that all mutable
-    memory returned by the call is not referenced by any other part of the
-    program, i.e. it is newly allocated by the function. Conversely, the mutable
-    references of the result may be assumed to not refer to any object that
-    existed before the function call. For example:)
+$(P A strongly pure function that accepts and returns by value can be assumed to return the same
+value for the same inputs. The compiler is allowed to memoize the result and reuse it for
+subsequent calls with identical arguments. Example:)
+
+---
+pure double fun(double);
+auto x = fun(1) + fun(1);
+// Code generated may be:
+// auto __t = fun(1);
+// auto x = __t + __t;
+---
+
+$(P For the purpose of deeming two calls to a strongly pure function equivalent, all arguments must
+be bitwise equal (as per the `is` operator), not logically equal (as per the `==` operator).
+Example:)
+
+$(SPEC_RUNNABLE_EXAMPLE_COMPILE
+---
+pure double fun(double x) { return x * 42; }
+pure int gun(int x) { return x ^ 42; }
+pure string hun(string s) { return s; }
+auto a = fun(1.0) + fun(1.0); // can memoize
+assert(-0.0 == 0.0 && -0.0 !is 0.0);
+auto b = fun(-0.0) + fun(0.0); // can NOT memoize, -0.0 and 0.0 have different bit patterns
+auto c = hun("ab") ~ hun("ab"); // can memoize ONLY if the two strings are interned
+---
+)
+
+$(P An implementation may not evaluate a strongly pure function speculatively. This is because a
+strongly pure function may abort the program, throw, or never terminate. Example:)
+
+$(SPEC_RUNNABLE_EXAMPLE_COMPILE
+---
+pure double fun(double x) { return x + 42; }
+pure double gun(size_t n)
+{
+    double result = 0;
+    foreach (i; 0 .. 100)
+    {
+        if (i == n) break;
+        result += fun(42);
+    }
+    return result;
+/*
+This code is NOT equivalent because it evaluates fun speculatively for n == 0:
+    double result = 0;
+    double __t = fun(42);
+    foreach (i; 0 .. 100)
+    {
+        if (i == n) break;
+        result += __t;
+    }
+    return result;
+*/
+}
+---
+)
+
+$(P A call to a `nothrow` strongly pure function that accepts and returns by value can be moved
+relative to its surrounding code, as long as its data dependencies are respected and there is no
+change in control flow. Example:)
+
+---
+pure nothrow double fun(double);
+pure nothrow double gun(double);
+void hun(double x, size_t n)
+{
+    auto y = fun(x) + gun(x); // evaluate fun(x) and gun(x) in any order
+    foreach (i; 0 .. 100)
+    {
+        // may hoist fun(10) outside the loop because it runs at least once
+         y += fun(10);
+    }
+    foreach (i; 0 .. n)
+    {
+        // may NOT hoist fun(42) outside the loop because it never run
+        // (i.e. counts as a speculative call)
+        y += fun(42);
```

Comment:
I'm not clear what's wrong with a speculative call to a nothrow function, assuming the 2 other points you used to reject speculative calls are agreed to be invalid.

Reply:
I think the existing stance is "if it returned x on the first call, it can be assumed it will return x on subsequent calls with the same arguments". So one or more calls can be folded into one. But not zero or more.

Reply:
>Key use-case being that speculative execution and memoisation would occur at compile-time...

This is not the charter of the PR at all..

Reply:
> I think the existing stance is "if it returned x on the first call, it can be assumed it will return x on subsequent calls with the same arguments". So one or more calls can be folded into one. But not zero or more.

Why? We agree that a `nothrow` pure function doesn't change flow-control, and the outputs are a map of the inputs... it doesn't matter if the number of calls is 0, 1, or more.

My read is the stance: "it returns X<sub>result</sub> when called with the arguments X<sub>inputs...</sub>". What are the cases where turning 0 evaluations into 1 is invalid?
If that's actually invalid, then I remain suspect of our definition of 'strongly pure'... there must be side-effects somewhere?

> This is not the charter of the PR at all..

I'm just pointing out an interesting case where speculative execution may be useful. I could imagine many more... perhaps the optimiser analyses the loop and notes that a call is made with one particular argument with very high probability and chooses to memoise that high-probability case?
Point being, I see no reason to spec it out of existence without good reason.

Reply:
Okay, thinking further, I guess it's this:
```d
pure float rcp(float x) { return 1/x; }

rcp(0);
```
We couldn't speculatively execute a runtime error.

Reply:
@TurkeyMan roger, though a better example would use `int` (`float` division by zero yields `inf`).

Reply:
I'm still not convinced either way though; in your case, you might as well rewrite the code to this:
```d
foreach (i; 0 .. n)
{
  crash();
}
```
Or rather
```d
if (n)
  crash();
```
So, we can't do a speculative evaluation because there is a hard-crash deliberately compiled into the code, just waiting for the runtime condition to trigger it?

It's true that such a runtime condition may not occur at runtime, and the speculative execution would be the only thing that triggers it... but it still feels weird to write the spec to specifically support the fact a manually (and deliberately) deployed runtime error is good enough reason to inhibit a class of optimisations.

I'd rather see that remain undefined rather than explicitly inhibited, and allow the judgement of the implementation to make that call; perhaps there's some implicit information available from value range propagation or something that assists the compiler in determining that such a speculative call will occur unconditionally, and then a speculative call becomes fine again.

Reply:
In this case, if the user wants to write `crash()`, then they should just make it impure (program termination is impure). And if they want to call `crash()` from inside a pure function, it's correct that it should be rejected, because crashing is not a pure operation.

I think your example is only valid if the function receives a non-constant argument. Your example receives a literal, and in that case, I think the compiler should reasonably be allowed to assume the call is valid, and it does not crash.

Reply:
I don't understand this... it's not important whether fun is called with a literal or a variable. The point is: "if fun(x) returns y once it can always be assumed that fun(x) returns y; it fun(x) throws once then fun(x) always throws; if fun(x) aborts then fun(x) always aborts; if fun(x) enters an infinite loop then fun(x) will always enter an infinite loop". That's why you can't do speculative calls (because you don't know what's going to happen).

Don't forget there's no access to fun's body.

To summarize: speculative execution is never allowed because there's no knowledge of the effect. But the result of the function is not dependent on time.

Reply:
And there's no compile-time aspect to this.

Reply:
> speculative execution is never allowed because there's no knowledge of the effect

And my point here is, I think it's reasonable to make assumptions about whether the function does those things you just said or not (I suggest it does not), and from there, we don't need to spec that it's invalid, and leave it to the implementation.

> if fun(x) returns y once it can always be assumed that fun(x) returns y;

Agree.

> if fun(x) throws once then fun(x) always throws;

Agree, and the `nothrow` case is interesting, that's what leads here.

> if fun(x) aborts then fun(x) always aborts;

Abort is absolutely not pure. If a pure function aborts, you've executed an invalid program.
I think the only valid case of 'abort' is runtime error due to invalid argument. Any attempt to deliberately abort violates functional purity, and we should emit any errors possible relating to this (ie, `exit()` is not a pure function).

This is the case I'm trying to address; I'm saying that, when called with a literal as in your example, the compiler may reasonably assume that such a call will not abort. The argument is not an unknown runtime argument requiring the compiler to be conservative about speculative execution potentially triggering a runtime error. If the user supplied an invalid argument as a literal, then they simply authored an invalid program, and I don't think we should add text to the spec to protect against this at the expense of opportunity for optimisation. I think it's better to leave it unspecified, and implementations can take advantage if they find value.

> if fun(x) enters an infinite loop then fun(x) will always enter an infinite loop

I think this is also a plainly invalid program, as I mentioned above. The compiler can assume this is not the case.

Reply:
> And there's no compile-time aspect to this.

This is a programming language that is processed by a compiler; everything that D has to say or do about anything has a compile-time aspect to it. Inhibiting the optimiser when we don't need to is important. The spec dictates what implementations may or may not do, and this line of text inhibits some potential optimisations. I want to be convinced this rule is definitely necessary, or remove the line form the spec if it's not.

Reply:
>Agree, and the nothrow case is interesting, that's what leads here.

It's not interesting because a nothrow function can still throw Error.

Reply:
>Abort is absolutely not pure.

You'll need to take that debate to another PR. This does not attempt to change the spec in that matter.

Reply:
>I'm saying that, when called with a literal as in your example, the compiler may reasonably assume that such a call will not abort.

How would it do that? The body of the function is not available for inspection. It is difficult for me what you need to get convinced of. Sorry.

Reply:
> It's not interesting because a nothrow function can still throw Error.

Are you saying a nothrow function can throw? What does nothrow mean then?

> How would it do that?

I've tried to show a number of times that, arguably, all of the things that need to be accounted for either violate purity, or are just plain invalid programs.
If it's the case that strong purity prevents the issues that you're trying to account for, then it's redundant.

Reply:
> You'll need to take that debate to another PR. This does not attempt to change the spec in that matter.

Where does the spec define abort as functionally pure? Is that a 'change'?

Reply:
>Are you saying a nothrow function can throw?

Affirmative.

>What does nothrow mean then?

Per the spec (incidentally this document): ["Nothrow functions can only throw exceptions derived from class Error."](https://dlang.org/spec/function.html#nothrow-functions)

>Where does the spec define abort as strongly pure? Is that a 'change'?

An application may abort execution by many means without calling `abort()`, e.g. through an integral division by zero, which is a pure operation.

A practical way of looking at things is that a pure function may have an arbitrarily high cost of execution. Therefore, speculatively calling it would not be advantageous. The preexisting spec's intention was to disallow speculative execution but allow memoization for calls with identical arguments. This PR does not attempt to change that; instead, it tries to better define things like what "identical arguments" really means, and how memoization can be effected. If there's better wording for that, please advise. Otherwise, I would appreciate if you took this debate outside this PR. It introduces its own controversy already and I don't think I can properly debate on too many fronts. Thanks.